server/src/main/java/org/elasticsearch/index/codec/tsdb/DocValuesForUtil.java

```diff
@@ -25,32 +29,82 @@ public DocValuesForUtil(int blockSize) {
+        if (bitsPerValue > 32 && bitsPerValue <= BITS_IN_FIVE_BYTES) {
+            return BITS_IN_FIVE_BYTES;
+        } else if (bitsPerValue > BITS_IN_FIVE_BYTES && bitsPerValue <= BITS_IN_SIX_BYTES) {
+            return BITS_IN_SIX_BYTES;
+        } else if (bitsPerValue > BITS_IN_SIX_BYTES && bitsPerValue <= BITS_IN_SEVEN_BYTES) {
+            return BITS_IN_SEVEN_BYTES;
+        }
+        return bitsPerValue;
+    }
+
+    void encode(long[] in, int bitsPerValue, final DataOutput out) throws IOException {
         if (bitsPerValue <= 24) { // these bpvs are handled efficiently by ForUtil
             forUtil.encode(in, bitsPerValue, out);
         } else if (bitsPerValue <= 32) {
             collapse32(in);
             for (int i = 0; i < blockSize / 2; ++i) {
                 out.writeLong(in[i]);
             }
+        } else if (bitsPerValue == BITS_IN_FIVE_BYTES || bitsPerValue == BITS_IN_SIX_BYTES || bitsPerValue == BITS_IN_SEVEN_BYTES) {
+            encodeFiveSixOrSevenBytesPerValue(in, bitsPerValue, out);
         } else {
+            assert bitsPerValue > 56 : "bitsPerValue must be greater than 56 but was [" + bitsPerValue + "]";
             for (long l : in) {
                 out.writeLong(l);
             }
         }
     }
 
-    void decode(int bitsPerValue, DataInput in, long[] out) throws IOException {
+    private void encodeFiveSixOrSevenBytesPerValue(long[] in, int bitsPerValue, final DataOutput out) throws IOException {
+        final DirectWriter writer = DirectWriter.getInstance(out, in.length, bitsPerValue);
+        for (long l : in) {
+            writer.add(l);
+        }
+        writer.finish();
+    }
+
+    void decode(int bitsPerValue, final DataInput in, long[] out) throws IOException {
         if (bitsPerValue <= 24) {
             forUtil.decode(bitsPerValue, in, out);
         } else if (bitsPerValue <= 32) {
             in.readLongs(out, 0, blockSize / 2);
             expand32(out);
+        } else if (bitsPerValue == BITS_IN_FIVE_BYTES || bitsPerValue == BITS_IN_SIX_BYTES || bitsPerValue == BITS_IN_SEVEN_BYTES) {
+            decodeFiveSixOrSevenBytesPerValue(bitsPerValue, in, out);
         } else {
+            assert bitsPerValue > 56 : "bitsPerValue must be greater than 56 but was [" + bitsPerValue + "]";
             in.readLongs(out, 0, blockSize);
         }
     }
 
+    private void decodeFiveSixOrSevenBytesPerValue(int bitsPerValue, final DataInput in, long[] out) throws IOException {
+        // NOTE: we expect multibyte values to be written "least significant byte" first
+        int bytesPerValue = bitsPerValue / Byte.SIZE;
+        for (int longValueIndex = 0; longValueIndex < blockSize; longValueIndex++) {
+            out[longValueIndex] = 0;
+            for (int byteIndex = 0; byteIndex < bytesPerValue; byteIndex++) {
+                byte b = in.readByte();
+                out[longValueIndex] += ((long) b & 0xFFL) << (8 * byteIndex);
+            }
```

Comment:
At first sight this decoding logic looks slow as I don't expect it to get vectorized. Did you compare it with something that is more similar to `ForUtil`'s logic?

Reply:
Maybe if I merge #92934 I can use the microbenchmark here and use those. I don't expect to be vectorized anyway.

Reply:
@jpountz do you think that a non-vectorized implementation might be too slow (at query time) for us to be able to use it?

Reply:
I'd just like to have something that is as fast as we can make it, given how changing file formats is a bit of a pain in general. So that we're not tempted to go with another layout in 1 month.